File: misc/surrounded_regions.py
from typing import List


class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """

        def mark_not_surrounded(r, c):
            board[r][c] = "N"

            for i, j in [(0, 1), (1, 0), (0, -1), (-1, 0)]:
                new_r, new_c = r + i, c + j

                if 0 <= new_r < len(board) and 0 <= new_c < len(board[0]) and \
                   board[new_r][new_c] == "O":
                   mark_not_surrounded(new_r, new_c)

        
        for r in range(len(board)):
            for c in range(len(board[0])):
                if board[r][c] == "O" and (
                    r == 0 or r == len(board) - 1 or c == 0 or c == len(board[0]) - 1
                ):
                    mark_not_surrounded(r, c)

        for r in range(len(board)):
            for c in range(len(board[0])):
                if board[r][c] == "O":
                    board[r][c] = "X"
                elif board[r][c] == "N":
                    board[r][c] = "O"

File: NeetCode_Roadmap/surrounded_regions.cpp
class Solution {
public:
    void mark_uncapturable(vector<vector<char>>& board, int r, int c)
    {
        board[r][c] = 'Z';
        for (auto [i, j] : {pair(0, 1), pair(1, 0), pair(0, -1), pair(-1, 0)})
        {
            int new_r = r + i;
            int new_c = c + j;
            if (0 <= new_r && new_r < board.size() && 0 <= new_c && new_c < board[0].size() &&
                board[new_r][new_c] == 'O')
            {
                mark_uncapturable(board, new_r, new_c);
            }
        }
    }

    void solve(vector<vector<char>>& board) {   
        // left + right
        for (int i = 0; i < board.size(); ++i)
        {
            // left
            if (board[i][0] == 'O') 
                mark_uncapturable(board, i, 0);
            // right
            if (board[i][board[0].size()-1] == 'O') 
                mark_uncapturable(board, i, board[0].size()-1);
        }

        // top + bottom
        for (int i = 0; i < board[0].size(); ++i)
        {
            // top
            if (board[0][i] == 'O')
                mark_uncapturable(board, 0, i);
            // bottom
            if (board[board.size()-1][i] == 'O')
                mark_uncapturable(board, board.size()-1, i);
        }

        for (int i = 0; i < board.size(); ++i)
        {
            for (int j = 0; j < board[0].size(); ++j)
            {
                if (board[i][j] == 'O') board[i][j] = 'X';
                if (board[i][j] == 'Z') board[i][j] = 'O';
            }
        }
    }
};